src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -219,3 +220,343 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
     fgMorphBlockStmt(block, stmt DEBUGARG(__FUNCTION__));
     return true;
 }
+
+//------------------------------------------------------------------------
+// optJumpThread: try and bypass the current block by rerouting
+//   flow from predecessors directly to successors.
+//
+// Arguments:
+//   block - block with branch to optimize
+//   domBlock - a dominating block that has an equivalent branch
+//
+// Returns:
+//   True if the branch was optimized.
+//
+// Notes:
+//
+//    A       B          A     B
+//     \     /           |     |
+//      \   /            |     |
+//      block     ==>    |     |
+//      /   \            |     |
+//     /     \           |     |
+//    C       D          C     D
+//
+bool Compiler::optJumpThread(BasicBlock* const block, BasicBlock* const domBlock)
+{
+    assert(block->bbJumpKind == BBJ_COND);
+    assert(domBlock->bbJumpKind == BBJ_COND);
+
+    // If the dominating block is not the immediate dominator
+    // we would need to duplicate a lot of code to thread
+    // the jumps. Pass for now.
+    //
+    if (domBlock != block->bbIDom)
+    {
+        return false;
+    }
+
+    JITDUMP("Both successors of IDom " FMT_BB " reach " FMT_BB " -- attemting jump threading\n", domBlock->bbNum,
+            block->bbNum);
+
+    // Since flow is going to bypass block, make sure there
+    // is nothing in block that can cause a side effect.
+    //
+    // Note we neglect PHI assignments. This reflects a general lack of
+    // SSA update abilities in the jit. We really should update any uses
+    // of PHIs defined here with the corresponding PHI input operand.
+    //
+    // TODO: if block has side effects, for those predecessors that are
+    // favorable (ones that don't reach block via a critical edge), consider
+    // duplicating block's IR into the predecessor. This is the jump threading
+    // analog of the optimization we encourage via fgOptimizeUncondBranchToSimpleCond.
+    //
+    Statement* const lastStmt = block->lastStmt();
+
+    for (Statement* stmt = block->FirstNonPhiDef(); stmt != nullptr; stmt = stmt->GetNextStmt())
+    {
+        GenTree* const tree = stmt->GetRootNode();
+
+        // We can ignore exception side effects in the jump tree.
+        //
+        // They are covered by the exception effects in the dominating compare.
+        // We know this because the VNs match and they encode exception states.
+        //
+        if ((tree->gtFlags & GTF_SIDE_EFFECT) != 0)
+        {
+            if (stmt == lastStmt)
+            {
+                assert(tree->OperIs(GT_JTRUE));
+                if ((tree->gtFlags & GTF_SIDE_EFFECT) == GTF_EXCEPT)
+                {
+                    // However, be conservative if block is in a try as we might not
+                    // have a full picture of EH flow.
+                    //
+                    if (!block->hasTryIndex())
```

Comment:
Good question.

I need to take a deeper look at how the jit models EH flow. It seems like some aspects are incomplete or approximate and so it might be dangerous to make assumptions when `fgReachable` returns false. I think we are safeguarded here and in the previous work I did because dominance information in the flow graph is also approximate in a way. But I am not 100% sure.

For example, in a case like:
```c#
    public static int Main(string[] args)
    {
        int x = args.Length;
        int y = 0;

        try 
        {
            if (x == 0)
            {
                throw new Exception();
            }
        }
        catch (Exception e)
        {
        }

        if (x == 0)
        {
            y = 100;
        }

        return y;
    }
```
We possibly could see that the second `(x == 0)` is dominated by the first `(x == 0)`, and if we did, `fgReachable` might only see the false path between the two, because the true path involves a throw/catch. If this happened we might mistakenly believe the second compare is always false. However, when we build dominators the second compare is not dominated by the first, so we never consider jump threading at all.
```
After computing reachability sets:
------------------------------------------------
BBnum  Reachable by 
------------------------------------------------
BB01 : BB01 
BB02 : BB01 BB02 
BB03 : BB01 BB02 BB03 
BB04 : BB01 BB02 BB04 BB07 
BB05 : BB01 BB02 BB04 BB05 BB07 
BB06 : BB01 BB02 BB04 BB05 BB06 BB07 
BB07 : BB07 

After computing reachability:

------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]      
------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1       [000..006)                                   
BB02 [0001]  1  0    BB01                  1       [006..009)-> BB04 ( cond ) T0      try {     
BB03 [0002]  1  0    BB02                  0       [009..00F)        (throw ) T0      }         
BB04 [0005]  2       BB02,BB07             1       [014..017)-> BB06 ( cond )                   
BB05 [0006]  1       BB04                  1       [017..01A)                                   
BB06 [0007]  2       BB04,BB05             1       [01A..01C)        (return)                   
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++        
BB07 [0004]  1     0                       1       [011..014)-> BB04 ( cret )    H0 F catch { } 
------------------------------------------------------------------------------------------------

*************** In fgComputeDoms

Dominator computation start blocks (those blocks with no incoming edges):
BB01 BB07 
------------------------------------------------
BBnum  Dominated by
------------------------------------------------
BB07:  BB07 
BB01:  BB01 
BB02:  BB02 BB01 
BB03:  BB03 BB02 BB01 
BB04:  BB04 
BB05:  BB05 BB04 
BB06:  BB06 BB04 

Inside fgBuildDomTree

After computing the Dominance Tree:
BB01 : BB02 
BB02 : BB03 
BB04 : BB06 BB05 
```
Note in the above that `BB04` is not dominated by `BB02`, because its pred `BB07` is an "eh entry". But what's odd is that neither `BB04` or `BB07` are dominated by `BB01` but reachability says they are reachable from `BB01`.

So mixing information from `fgComputeDoms` and `fgReachable` may be risky.

I can't come up with an example with the current opts where we'd make a bad assumption, but if/ when we extend things to handle "related" compares then perhaps we'll run into trouble.

What's also interesting is that SSA has its own dominance computation, and does it differently and seemingly a bit more accurately; here `BB01` properly dominates `BB04` and `BB07`. Seems like we ought not to have two different notions of dominance in the jit.
```
*************** In SsaBuilder::Build()
[SsaBuilder] Max block count is 8.

------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]      
------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1       [000..006)                                   
BB02 [0001]  1  0    BB01                  1       [006..009)-> BB04 ( cond ) T0      try {     
BB03 [0002]  1  0    BB02                  0       [009..00F)        (throw ) T0      }         
BB04 [0005]  2       BB02,BB07             1       [014..017)-> BB06 ( cond )                   
BB05 [0006]  1       BB04                  1       [017..01A)                                   
BB06 [0007]  2       BB04,BB05             1       [01A..01C)        (return)                   
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
BB07 [0004]  1     0                       0       [011..014)-> BB04 ( cret )    H0 F catch { } 
------------------------------------------------------------------------------------------------

***************  Exception Handling table
index  eTry, eHnd
  0  ::            - Try at BB02..BB03 [006..011), Handler at BB07..BB07 [011..014)
[SsaBuilder] Topologically sorted the graph.
[SsaBuilder::ComputeImmediateDom]

Inside fgBuildDomTree

After computing the Dominance Tree:
BB01 : BB07 BB04 BB02 
BB02 : BB03 
BB04 : BB06 BB05 
```

I'm not going to have time to unravel this before the holidays, so will come back to it in the new year.

Reply:
Am going to add a bespoke reachability computation (`optReachable`) for this.